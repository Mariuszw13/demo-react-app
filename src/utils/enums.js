export const ITEM_TYPE = {
    FOOD: "food",
    DRINK: "drink"
};