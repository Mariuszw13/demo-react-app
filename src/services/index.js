import axios from "axios";
import { noop } from "../utils";

const instance = axios.create({ baseURL: process.env.REACT_APP_API_URL });

export const tryLogin = async (credentials, onSuccess, onFailure = noop) => {
    try {
        const response = await instance.get("/auth", credentials);
        const token = response.data.token;
        onSuccess(token);
    } catch (e) {
        console.log(e);
        onFailure();
    }
};

export const getItems = async (onSuccess, token, onFailure = noop) => {
    try {
        const config = { headers: { "X-Token": token } };
        const response = await instance.get("/items", config);
        onSuccess(response.data);
    } catch (e) {
        console.log(e);
        onFailure();
    }
};

export const deleteItem = async (onSuccess, token, id, onFailure = noop) => {
    try {
        const config = { headers: { "X-Token": token } };
        const response = await instance.delete(`/items/${id}`, config);
        onSuccess(response.data);
    } catch (e) {
        console.log(e);
        onFailure();
    }
};

export const addItem = async (onSuccess, token, newItem, onFailure = noop) => {
    try {
        const config = { headers: { "X-Token": token } };
        const response = await instance.post("/items", newItem, config);
        onSuccess(response.data);
    } catch (e) {
        console.log(e);
        onFailure();
    }
};
