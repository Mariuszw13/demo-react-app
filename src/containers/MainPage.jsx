import React, {useState, useEffect} from 'react';
import {addItem, deleteItem, getItems} from "../services";
import {useCookies} from "react-cookie";
import List from "../components/item/list";
import styled from "styled-components";
import NewItemModal from "../components/item/form/NewItemModal";
import Button from "react-bootstrap/Button";

const MainPage = ({className}) => {
    const [items, setItems] = useState([]);
    const [modalVisible, setModalVisibility] = useState(false);
    const [cookies,] = useCookies(['authToken']);

    useEffect(() => {
        getItems(setItems, cookies.authToken);
    }, [cookies]);

    const removeItem = id => {
        deleteItem(onDeleteSuccess(id), cookies.authToken, id);
    };

    const addNewItem = newItem => {
        addItem(onAddSuccess(newItem), cookies.authToken, newItem);
    };

    const onDeleteSuccess = id => () => {
        const newArr = [...items];
        const index = newArr.findIndex(item => item.id === id);
        if (index > -1) {
            newArr.splice(index, 1);
            setItems(newArr)
        }
    };

    const onAddSuccess = newItem => () => {
        const newArr = [...items, newItem];
        setItems(newArr)
    };

    const openModal = () => setModalVisibility(true);
    const closeModal = () => setModalVisibility(false);

    return (
        <div className={className}>
            <Button onClick={openModal}>Add new item</Button>
            <List items={items} removeItem={removeItem}/>
            <NewItemModal visible={modalVisible} closeModal={closeModal} addNewItem={addNewItem}/>
        </div>
    )
};

export default styled(MainPage)`
    padding: 10px 10%;
`;