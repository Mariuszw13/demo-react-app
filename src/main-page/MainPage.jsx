import React, { useState, useEffect } from "react";
import { addItem, deleteItem, getItems } from "../services";
import List from "../components/item/list";
import styled from "styled-components";
import NewItemModal from "../components/item/form/NewItemModal";
import Button from "react-bootstrap/Button";
import { useAuthToken } from "../login/useAuthToken";
import * as PropTypes from "prop-types";

const MainPage = ({ className }) => {
    const [items, setItems] = useState([]);
    const [lastAddedIndex, setLastAddedIndex] = useState(6);
    const [modalVisible, setModalVisibility] = useState(false);
    const authToken = useAuthToken();

    useEffect(() => {
        getItems(setItems, authToken);
    }, [authToken]);

    const removeItem = id => {
        // due to mocked api, there are 5 mocked items
        if (id < 6) {
            deleteItem(onDeleteSuccess(id), authToken, id);
        } else {
            onDeleteSuccess(id)();
        }
    };

    const addNewItem = newItem => {
        addItem(onAddSuccess(newItem), authToken, newItem);
    };

    const onDeleteSuccess = id => () => {
        const newArr = items.filter(item => item.id !== id);
        setItems(newArr);
    };

    const onAddSuccess = newItem => () => {
        const newArr = [...items, newItem];
        setLastAddedIndex(newItem.id);
        setItems(newArr);
    };

    const openModal = () => setModalVisibility(true);
    const closeModal = () => setModalVisibility(false);
    
    return (
        <div className={className}>
            <Button onClick={openModal}>Add new item</Button>
            <List items={items} removeItem={removeItem} />
            <NewItemModal visible={modalVisible} closeModal={closeModal} addNewItem={addNewItem} lastAddedIndex={lastAddedIndex} />
        </div>
    );
};

MainPage.propTypes = {
    className: PropTypes.string,
};

export default styled(MainPage)`
    padding: 10px 10%;
`;
