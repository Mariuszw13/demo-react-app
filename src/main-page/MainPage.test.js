import React from "react";
import { render, waitFor } from "@testing-library/react";
import MainPage from "./MainPage";
import * as services from "../services";
import { ITEM_TYPE } from "../utils/enums";

describe("<MainPage />", () => {
    const addItemText = "Add new item";
    const data = [
        { id: 1, name: "item one", type: ITEM_TYPE.FOOD },
        { id: 2, name: "item two", type: ITEM_TYPE.DRINK },
    ];

    jest.spyOn(services, "getItems").mockImplementation((onSuccess) => onSuccess(data));
    jest.spyOn(services, "deleteItem").mockImplementation((onSuccess) => onSuccess());

    it("should render properly and match snapshot", async () => {
        // given
        const { container, getByText } = render(<MainPage />);
        // when
        await waitFor(() => getByText(addItemText));
        // then
        expect(container.firstChild).toMatchSnapshot();
    });

    it("should open modal on button click", async () => {
        // given
        const { getByText, getAllByText } = render(<MainPage />);
        // when
        const button = await waitFor(() => getByText(addItemText));
        button.click();
        // then
        expect(getAllByText(addItemText)).toHaveLength(2);
    });

    it("should render items fetched from api", async () => {
        const { getByText } = render(<MainPage />);
        // when
        await waitFor(() => getByText(addItemText));
        const item1 = getByText(data[0].name);
        const item2 = getByText(data[1].name);
        // then
        expect(item1).toBeInTheDocument();
        expect(item2).toBeInTheDocument();
    });

    it("should remove item on delete button click", async () => {
        // given
        const { queryByText, findAllByText } = render(<MainPage />);
        // when
        const deleteButtons = await findAllByText("Remove");
        const item1DeleteButton = deleteButtons[0];
        item1DeleteButton.click();
        // then
        await waitFor(() => {
            expect(queryByText(data[0].name)).not.toBeInTheDocument()
        })
    })
});
