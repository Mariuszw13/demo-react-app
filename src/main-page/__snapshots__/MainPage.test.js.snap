// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> should render properly and match snapshot 1`] = `
<div
  class="sc-AxhCb duPdoD"
>
  <button
    class="btn btn-primary"
    type="button"
  >
    Add new item
  </button>
  <div
    class="card-columns"
  >
    <div
      class="card"
    >
      <img
        class="card-img-top"
        src="https://picsum.photos/id/21/400"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          item one
        </div>
        <p
          class="card-text"
        />
        <div
          class="sc-AxiKw iSgsng"
        >
          <div
            class="sc-AxirZ ktRBae"
          >
            <span
              class="sc-AxjAm eNShYM"
            >
              Weight
            </span>
            <span>
              undefined kg
            </span>
          </div>
          <div
            class="sc-AxirZ ktRBae"
          >
            <span
              class="sc-AxjAm eNShYM"
            >
              Price
            </span>
            <span>
              undefined $
            </span>
          </div>
          <div
            class="sc-AxirZ ktRBae"
          >
            <span
              class="sc-AxjAm eNShYM"
            >
              Quantity
            </span>
            <span>
              undefined 
            </span>
          </div>
        </div>
        <button
          class="btn btn-danger"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
    <div
      class="card"
    >
      <img
        class="card-img-top"
        src="https://picsum.photos/id/22/400"
      />
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          item two
        </div>
        <div
          class="sc-AxiKw iSgsng"
        >
          <div
            class="sc-AxirZ ktRBae"
          >
            <span
              class="sc-AxjAm eNShYM"
            >
              Volume
            </span>
            <span>
              undefined l
            </span>
          </div>
          <div
            class="sc-AxirZ ktRBae"
          >
            <span
              class="sc-AxjAm eNShYM"
            >
              Price
            </span>
            <span>
              undefined $
            </span>
          </div>
          <div
            class="sc-AxirZ ktRBae"
          >
            <span
              class="sc-AxjAm eNShYM"
            >
              Quantity
            </span>
            <span>
              undefined 
            </span>
          </div>
        </div>
        <button
          class="btn btn-danger"
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
</div>
`;
